src/ppm.rs

```diff
@@ -0,0 +1,499 @@
+    let joint_rand: Vec<F> = (0..input.joint_rand_len())
+        .map(|_| prng.next().unwrap())
+        .collect();
+    let prove_rand: Vec<F> = rand(input.prove_rand_len())?;
+    let proof = prove(input, &prove_rand, &joint_rand)?;
+
+    // Generate the proof shares and finalize the joint randomness seed hints.
+    let proof_len = proof.as_slice().len();
+    let mut leader_proof_share = proof.data;
+    for helper in helper_shares.iter_mut() {
+        getrandom(&mut helper.proof_share)?;
+
+        let prng = Prng::try_from_seed(&helper.proof_share)?;
+        for (x, y) in leader_proof_share.iter_mut().zip(prng).take(proof_len) {
+            *x -= y;
+        }
+
+        for (x, y) in helper
+            .joint_rand_seed_hint
+            .iter_mut()
+            .zip(joint_rand_seed.iter())
+        {
+            *x ^= y;
+        }
+    }
+
+    for (x, y) in leader_joint_rand_seed_hint
+        .iter_mut()
+        .zip(joint_rand_seed.iter())
+    {
+        *x ^= y;
+    }
+
+    // Prepare the output messages.
+    let mut msgs = Vec::with_capacity(num_shares);
+    msgs.push(Upload {
+        input_share: Share::Leader(leader_input_share),
+        proof_share: Share::Leader(leader_proof_share),
+        joint_rand_seed_hint: leader_joint_rand_seed_hint,
+        blind: leader_blind,
+    });
+
+    for helper in helper_shares.into_iter() {
+        msgs.push(Upload {
+            input_share: Share::Helper(helper.input_share, input_len),
+            proof_share: Share::Helper(helper.proof_share, proof_len),
+            joint_rand_seed_hint: helper.joint_rand_seed_hint,
+            blind: helper.blind,
+        });
+    }
+
+    Ok(msgs)
+}
+
+/// The message sent by an aggregator to every other aggregator. This is the final message of the
+/// input-validation protocol.
+#[derive(Clone, Debug)]
+pub struct Verifier<F: FieldElement> {
+    /// The aggregator's share of the verifier message.
+    pub verifier_share: Vec<F>,
+
+    /// The aggregator's share of the joint randomness, derived from its input share.
+    pub joint_rand_seed_share: Vec<u8>,
+}
+
+/// The initial state of an aggregate upon receiving the [`Upload`] message from the client.
+#[derive(Clone, Debug)]
+pub struct AggregatorState<F, V>
+where
+    F: FieldElement,
+    V: pcp::Value<F>,
+{
+    input_share: V,
+    verifier_share: Vec<F>,
+    joint_rand_seed_hint: Vec<u8>,
+}
+
+/// Run by each aggregator, this consumes the [`Upload`] message sent from the client and
+/// produces the [`Verifier`] message that will be broadcast to the other aggregators. It also
+/// returns the aggregator's state.
+///
+/// # Parameters
+///
+/// * `msg` is the message sent by the client. It contains, among other things, an input and proof
+/// share.
+/// * `param` is used to reconstruct the input share from the raw message.
+/// * `query_rand_seed` is used to derive the query randomness shared by all the aggregators.
+pub fn verify_start<F, V>(
+    msg: Upload<F>,
```

Comment:
```suggestion
    upload_message: Upload<F>,
```

Reply:
I have a weak preference for a short and generic name for messages?

Reply:
It's clear enough as a local variable name,